File: AgoraChat/Code/Controllers/Chats/AgoraChatUserDataModel.m
//
//  AgoraChatUserDataModel.m
//  EaseIM
//
//  Created by zhangchong on 2020/12/3.
//  Copyright © 2020 zhangchong. All rights reserved.
//

#import "AgoraChatUserDataModel.h"

@implementation AgoraChatUserDataModel

- (instancetype)initWithUserInfo:(AgoraChatUserInfo *)userInfo
{
    if (self = [super init]) {
        _easeId = userInfo.userId;
        _showName = userInfo.nickName;
        _avatarURL = userInfo.avatarUrl;
        _defaultAvatar = [self getAvatar:userInfo.userId];
    }
    return self;
}

- (UIImage *)getAvatar:(NSString *)uName
{
    if ([uName isEqualToString:AgoraChatClient.sharedClient.currentUsername]) {
        NSUserDefaults *userDefault = [NSUserDefaults standardUserDefaults];
        NSString *imageName = [userDefault valueForKey:[NSString stringWithFormat:@"%@_avatar",AgoraChatClient.sharedClient.currentUsername]];
        return ImageWithName(imageName);
    } else {
        UIImage *originImage = nil;
        
        NSUserDefaults *userDefault = [NSUserDefaults standardUserDefaults];
        NSString *imageName = [userDefault objectForKey:uName];
        if (imageName && imageName.length > 0) {
            originImage = ImageWithName(imageName);
        } else {
            int random = arc4random() % 7 + 1;
            NSString *imgName = [NSString stringWithFormat:@"defatult_avatar_%@",@(random)];
            [userDefault setObject:imgName forKey:uName];
            originImage = ImageWithName(imgName);
            [userDefault synchronize];
        }
        
        return [originImage acd_scaleToAssignSize:CGSizeMake(kAvatarHeight, kAvatarHeight)];
    }
    
    return ImageWithName(@"defaultAvatar");
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Chats/AgoraChatUserDataModel.swift
//
//  AgoraChatUserDataModel.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/25.
//

import UIKit

class AgoraChatUserDataModel: NSObject, EaseUserProfile {
    var easeId: String
    var showName: String
    var avatarURL: String
    var defaultAvatar: UIImage
    
    init(userInfo: AgoraChatUserInfo) {
        self.easeId = userInfo.userId ?? ""
        self.showName = userInfo.showName
        self.avatarURL = userInfo.avatarUrl ?? ""
        
        var defaultAvatar: UIImage!
        if self.easeId == AgoraChatClient.shared().currentUsername {
            if let imageName = UserDefaults.standard.value(forKey: "\(self.easeId)_avatar") as? String, let image = UIImage(named: imageName) {
                defaultAvatar = image
            }
        } else {
            if let imageName = UserDefaults.standard.value(forKey: self.easeId) as? String, let image = UIImage(named: imageName) {
                defaultAvatar = image
            } else {
                let imageName = "defatult_avatar_\(arc4random() % 7 + 1)"
                UserDefaults.standard.set(imageName, forKey: self.easeId)
                UserDefaults.standard.synchronize()
                if let image = UIImage(named: imageName) {
                    defaultAvatar = image
                }
            }
        }
        if defaultAvatar == nil {
            defaultAvatar = UIImage(named: "defatult_avatar_1")!
        }
        self.defaultAvatar = defaultAvatar
        super.init()
    }
    
    private func getAvatar(userId: String) -> UIImage {
        if userId == AgoraChatClient.shared().currentUsername {
            if let imageName = UserDefaults.standard.value(forKey: "\(userId)_avatar") as? String, let image = UIImage(named: imageName) {
                return image
            }
        } else {
            if let imageName = UserDefaults.standard.value(forKey: userId) as? String, let image = UIImage(named: imageName) {
                return image
            } else {
                let imageName = "defatult_avatar_\(arc4random() % 7 + 1)"
                UserDefaults.standard.set(imageName, forKey: userId)
                UserDefaults.standard.synchronize()
                if let image = UIImage(named: imageName) {
                    return image
                }
            }
        }
        return UIImage(named: "defaultAvatar")!
    }
}
